File: output/test_file_generator/complete _purchase.java
import org.openqa.selenium.*;
import org.openqa.selenium.chrome.ChromeDriver;
import org.openqa.selenium.chrome.ChromeOptions;
import org.openqa.selenium.support.ui.WebDriverWait;
import org.openqa.selenium.support.ui.ExpectedConditions;
import org.testng.annotations.*;
import org.testng.Assert;
import io.qameta.allure.Allure;
import io.qameta.allure.Step;
import io.github.bonigarcia.wdm.WebDriverManager;
import java.io.ByteArrayInputStream;
import java.time.Duration;
import java.util.List;
import java.util.ArrayList;

public class TC01_Successful_purchase_standard_user {
    private WebDriver driver;
    private WebDriverWait wait;
    private final String baseUrl = "https://www.saucedemo.com/";

    @BeforeClass
    public void setUp() {
        WebDriverManager.chromedriver().setup();
        ChromeOptions options = new ChromeOptions();
        options.addArguments("--start-maximized");
        // optionally run headless in CI by uncommenting:
        // options.addArguments("--headless=new");
        driver = new ChromeDriver(options);
        wait = new WebDriverWait(driver, Duration.ofSeconds(15));
        Allure.step("Browser started and WebDriver initialized");
    }

    @AfterClass
    public void tearDown() {
        try {
            Allure.step("Closing browser");
        } catch (Exception ignored) {
        }
        if (driver != null) {
            driver.quit();
        }
    }

    @Test(description = "TC01 - Successful purchase (standard_user)")
    public void TC01_Successful_purchase_standard_user() {
        try {
            // 1) Open base URL
            Allure.step("Navigate to base URL: " + baseUrl);
            driver.get(baseUrl);
            attachScreenshot("after_navigate_base");

            // If a new window was expected in recording, attempt to switch to it
            switchToNewWindowIfAny();

            // 2) Login - enter username
            helper_sendKeys_withHealing("user-name", "standard_user", "Enter username (user-name)");
            // 3) Login - enter password
            helper_sendKeys_withHealing("password", "secret_sauce", "Enter password (password)");
            // 4) Click login
            helper_click_withHealing("login-button", "Click login button (login-button)");

            // 5) Add to cart - backpack
            helper_click_withHealing("add-to-cart-sauce-labs-backpack", "Add Sauce Labs Backpack to cart");

            // 6) Click cart
            helper_click_possibleLocators(new String[]{"shopping_cart_container", "cart", "shopping_cart_link"}, "Open cart");

            // 7) Click checkout
            helper_click_withHealing("checkout", "Click checkout");

            // 8) Fill first name, last name, postal code
            helper_sendKeys_withHealing("firstName", "test", "Enter first name");
            helper_sendKeys_withHealing("lastName", "test", "Enter last name");
            helper_sendKeys_withHealing("postalCode", "3242354", "Enter postal code");

            // 9) Continue
            helper_click_withHealing("continue", "Click continue");

            // 10) Finish
            helper_click_withHealing("finish", "Click finish to complete purchase");

            // Primary assertion: verify checkout complete page reached
            Allure.step("Primary assertion: verify checkout-complete page is reached");
            attachScreenshot("before_primary_assertion");
            try {
                wait.until(driver -> driver.getCurrentUrl() != null && driver.getCurrentUrl().contains("checkout-complete.html"));
                String currentUrl = driver.getCurrentUrl();
                Assert.assertTrue(currentUrl != null && currentUrl.contains("checkout-complete.html"),
                        "Expected to be on checkout-complete page but current URL was: " + currentUrl);
            } catch (Exception ae) {
                attachPageSource("assertion_failure_page_source");
                attachScreenshot("assertion_failure_screenshot");
                throw ae;
            }

            // 11) After finish click back-to-products
            helper_click_withHealing("back-to-products", "Click back to products");

            // 12) Open menu and logout
            helper_click_withHealing("react-burger-menu-btn", "Open side menu");
            // Click All Items if present (inventory_sidebar_link), then logout
            helper_click_possibleLocators(new String[]{"inventory_sidebar_link"}, "Click All Items (inventory_sidebar_link)");
            helper_click_withHealing("logout_sidebar_link", "Click logout");

            Allure.step("Test flow completed");
        } catch (Exception e) {
            attachPageSource("error_page_source");
            attachScreenshot("error_screenshot");
            Allure.step("Exception during test execution: " + e.getMessage());
            Assert.fail("Test encountered exception: " + e.getMessage());
        }
    }

    // Helper to click by id with healing algorithm and JS fallback
    @Step("{description}")
    private void helper_click_withHealing(String id, String description) {
        Allure.step("Attempting click for: " + id + " - " + description);
        Exception lastEx = null;

        // Try multiple strategies in order
        By[] primaryLocators = new By[]{
                By.id(id),
                By.name(id),
                By.xpath("//*[text()='" + id + "']"),
                By.xpath("//*[contains(@id, '" + id + "')]"),
                By.xpath("//button[contains(., '" + id + "')] | //a[contains(., '" + id + "')]")
        };

        for (By locator : primaryLocators) {
            try {
                wait.until(ExpectedConditions.presenceOfElementLocated(locator));
                wait.until(ExpectedConditions.visibilityOfElementLocated(locator));
                try {
                    wait.until(ExpectedConditions.elementToBeClickable(locator));
                } catch (Exception ignore) {
                    // element might be visible but not clickable; we'll attempt click regardless
                }
                WebElement el = driver.findElement(locator);
                try {
                    el.click();
                    attachScreenshot("click_success_" + locator.toString());
                    return;
                } catch (Exception clickEx) {
                    lastEx = clickEx;
                    Allure.step("Native click failed for locator " + locator + ": " + clickEx.getMessage());
                    // Try JS click as fallback
                    try {
                        ((JavascriptExecutor) driver).executeScript("arguments[0].click();", el);
                        attachScreenshot("jsclick_success_" + locator.toString());
                        Allure.step("JS click succeeded for locator " + locator);
                        return;
                    } catch (Exception jsEx) {
                        lastEx = jsEx;
                        Allure.step("JS click also failed for locator " + locator + ": " + jsEx.getMessage());
                    }
                }
            } catch (Exception e) {
                lastEx = e;
                Allure.step("Locator attempt failed: " + locator + " -> " + e.getMessage());
                attachScreenshot("locator_attempt_failed_" + locator.toString());
                sleep(300);
            }
        }

        // Final attempt: try clickable by CSS class matching common patterns
        try {
            By fallback = By.cssSelector("button, a, div");
            List<WebElement> els = driver.findElements(fallback);
            for (WebElement el : els) {
                try {
                    String txt = "";
                    try { txt = el.getText(); } catch (Exception ignored) {}
                    String attrId = "";
                    try { attrId = el.getAttribute("id"); } catch (Exception ignored) {}
                    if ((txt != null && txt.contains(id)) || (attrId != null && attrId.contains(id))) {
                        try {
                            el.click();
                            attachScreenshot("click_success_fallback_textmatch");
                            Allure.step("Clicked fallback element matching text/id: " + id);
                            return;
                        } catch (Exception e) {
                            try {
                                ((JavascriptExecutor) driver).executeScript("arguments[0].click();", el);
                                attachScreenshot("jsclick_success_fallback_textmatch");
                                Allure.step("JS clicked fallback element matching text/id: " + id);
                                return;
                            } catch (Exception inner) {
                                lastEx = inner;
                            }
                        }
                    }
                } catch (StaleElementReferenceException sere) {
                    // ignore stale and continue
                }
            }
        } catch (Exception e) {
            lastEx = e;
        }

        Allure.step("All healing attempts failed for '" + id + "'. Throwing exception.");
        if (lastEx != null) {
            throw new RuntimeException("Click action failed for '" + id + "': " + lastEx.getMessage(), lastEx);
        } else {
            throw new RuntimeException("Click action failed for '" + id + "': unknown reason");
        }
    }

    // Helper to send keys by id with healing and JS fallback
    @Step("{description}")
    private void helper_sendKeys_withHealing(String id, String keys, String description) {
        Allure.step("Attempting sendKeys for: " + id + " - " + description);
        Exception lastEx = null;

        By[] locators = new By[]{
                By.id(id),
                By.name(id),
                By.xpath("//input[@placeholder='" + id + "']"),
                By.xpath("//input[contains(@id, '" + id + "')]"),
                By.xpath("//input")
        };

        for (By locator : locators) {
            try {
                wait.until(ExpectedConditions.presenceOfElementLocated(locator));
                wait.until(ExpectedConditions.visibilityOfElementLocated(locator));
                WebElement el = driver.findElement(locator);
                try {
                    el.clear();
                } catch (Exception ignore) {
                }
                try {
                    el.sendKeys(keys);
                    attachScreenshot("sendkeys_success_" + locator.toString());
                    return;
                } catch (Exception sendEx) {
                    lastEx = sendEx;
                    Allure.step("Native sendKeys failed for locator " + locator + ": " + sendEx.getMessage());
                    // Try JS set value
                    try {
                        ((JavascriptExecutor) driver).executeScript("arguments[0].value = arguments[1]; arguments[0].dispatchEvent(new Event('input'));", el, keys);
                        attachScreenshot("js_sendkeys_success_" + locator.toString());
                        Allure.step("JS set value succeeded for locator " + locator);
                        return;
                    } catch (Exception jsEx) {
                        lastEx = jsEx;
                        Allure.step("JS set value failed for locator " + locator + ": " + jsEx.getMessage());
                    }
                }
            } catch (Exception e) {
                lastEx = e;
                Allure.step("Locator attempt failed for sendKeys: " + locator + " -> " + e.getMessage());
                attachScreenshot("sendkeys_locator_attempt_failed_" + locator.toString());
                sleep(300);
            }
        }

        Allure.step("All healing attempts failed for sendKeys for '" + id + "'. Throwing exception.");
        if (lastEx != null) {
            throw new RuntimeException("sendKeys action failed for '" + id + "': " + lastEx.getMessage(), lastEx);
        } else {
            throw new RuntimeException("sendKeys action failed for '" + id + "': unknown reason");
        }
    }

    // Helper to click using multiple possible ids (useful when element id varies)
    @Step("{description}")
    private void helper_click_possibleLocators(String[] ids, String description) {
        for (String id : ids) {
            try {
                helper_click_withHealing(id, description + " (attempting id: " + id + ")");
                return;
            } catch (Exception e) {
                Allure.step("Attempt for locator '" + id + "' failed: " + e.getMessage());
                // continue to next
            }
        }
        // As final fallback, try a generic cart icon or button
        try {
            By cartBy = By.xpath("//a[contains(@class,'shopping_cart_link')] | //div[@id='shopping_cart_container'] | //a[@id='shopping_cart_container']");
            wait.until(ExpectedConditions.presenceOfElementLocated(cartBy));
            WebElement cartBtn = driver.findElement(cartBy);
            try {
                cartBtn.click();
            } catch (Exception e) {
                ((JavascriptExecutor) driver).executeScript("arguments[0].click();", cartBtn);
            }
            attachScreenshot("click_success_generic_cart");
            Allure.step("Healed by clicking generic cart selector");
            return;
        } catch (Exception e) {
            attachScreenshot("click_generic_cart_failed");
            throw new RuntimeException("All attempts failed for clicking one of possible locators", e);
        }
    }

    @Step("Switch to new window if any")
    private void switchToNewWindowIfAny() {
        Allure.step("Checking for a new window to switch to");
        try {
            String original = driver.getWindowHandle();
            List<String> handles = new ArrayList<>(driver.getWindowHandles());
            if (handles.size() > 1) {
                for (String h : handles) {
                    if (!h.equals(original)) {
                        driver.switchTo().window(h);
                        Allure.step("Switched to new window: " + driver.getCurrentUrl());
                        attachScreenshot("switched_to_new_window");
                        return;
                    }
                }
            } else {
                // Poll for a short time
                long start = System.currentTimeMillis();
                while ((System.currentTimeMillis() - start) < 5000) {
                    handles = new ArrayList<>(driver.getWindowHandles());
                    if (handles.size() > 1) {
                        for (String h : handles) {
                            if (!h.equals(original)) {
                                driver.switchTo().window(h);
                                Allure.step("Switched to new window after wait: " + driver.getCurrentUrl());
                                attachScreenshot("switched_to_new_window");
                                return;
                            }
                        }
                    }
                    sleep(200);
                }
                Allure.step("No additional window found to switch to");
            }
        } catch (Exception e) {
            Allure.step("Exception while switching to new window: " + e.getMessage());
            attachScreenshot("switch_window_exception");
        }
    }

    @Step("Attach screenshot: {name}")
    private void attachScreenshot(String name) {
        try {
            Allure.addAttachment(name, new ByteArrayInputStream(((TakesScreenshot) driver).getScreenshotAs(OutputType.BYTES)));
        } catch (Exception e) {
            // best-effort only
        }
    }

    @Step("Attach page source: {name}")
    private void attachPageSource(String name) {
        try {
            String src = driver.getPageSource();
            Allure.addAttachment(name + ".html", new ByteArrayInputStream(src.getBytes()));
        } catch (Exception e) {
            // best effort
        }
    }

    private void sleep(long ms) {
        try {
            Thread.sleep(ms);
        } catch (InterruptedException ignored) {
        }
    }
}

File: output/test_file_generator/complete_purchase.py
import os
import time
import json
import pytest
import allure
from selenium import webdriver
from selenium.webdriver.common.by import By
from selenium.webdriver.chrome.options import Options
from selenium.webdriver.support.ui import WebDriverWait
from selenium.webdriver.support import expected_conditions as EC
from allure_commons.types import AttachmentType

BASE_URL = "https://www.saucedemo.com/"

PAGE_FILE_KEY_SN = "jan_sauce_demo_login_py"


def ensure_reports_dir():
    reports_dir = os.path.join(os.getcwd(), "reports")
    os.makedirs(reports_dir, exist_ok=True)
    return reports_dir


def switch_to_new_window(driver, timeout=10):
    start = time.time()
    while time.time() - start < timeout:
        if len(driver.window_handles) > 1:
            driver.switch_to.window(driver.window_handles[-1])
            return
        time.sleep(0.5)
    raise RuntimeError("New window did not appear within timeout")


def _attach_screenshot(driver, name="screenshot"):
    try:
        allure.attach(driver.get_screenshot_as_png(), name=name, attachment_type=AttachmentType.PNG)
    except Exception:
        # best-effort
        pass


def _write_report_artifacts(driver, tag="failure"):
    reports_dir = ensure_reports_dir()
    ts = int(time.time() * 1000)
    safe_name = f"{tag}_{ts}"
    try:
        png_path = os.path.join(reports_dir, f"{safe_name}.png")
        with open(png_path, "wb") as f:
            f.write(driver.get_screenshot_as_png())
    except Exception:
        pass
    try:
        html_path = os.path.join(reports_dir, f"{safe_name}.html")
        with open(html_path, "w", encoding="utf-8") as f:
            f.write(driver.page_source)
    except Exception:
        pass
    try:
        meta = {"url": driver.current_url, "timestamp": ts}
        json_path = os.path.join(reports_dir, f"{safe_name}.json")
        with open(json_path, "w", encoding="utf-8") as f:
            json.dump(meta, f)
    except Exception:
        pass
    return reports_dir


def find_element_with_healing(driver, locator_candidates, desc, timeout=5):
    # locator_candidates: list of (By, selector) tuples
    last_exception = None
    with allure.step(f"Find element: {desc} using candidates: {locator_candidates}"):
        _attach_screenshot(driver, name=f"before_find_{desc}")
        # 1) Try direct attempts quickly
        for by, sel in locator_candidates:
            try:
                el = driver.find_element(by, sel)
                _attach_screenshot(driver, name=f"found_direct_{desc}")
                return el
            except Exception as e:
                last_exception = e
                continue
        # 2) Retry once after a short wait (replacing sleep with a short WebDriverWait)
        try:
            WebDriverWait(driver, 0.6).until(lambda d: True)
        except Exception:
            pass
        for by, sel in locator_candidates:
            try:
                el = driver.find_element(by, sel)
                _attach_screenshot(driver, name=f"found_retry_{desc}")
                return el
            except Exception as e:
                last_exception = e
                continue
        # 3) Explicit wait fallback per locator candidates
        for by, sel in locator_candidates:
            try:
                el = WebDriverWait(driver, timeout).until(EC.visibility_of_element_located((by, sel)))
                _attach_screenshot(driver, name=f"found_wait_{desc}")
                return el
            except Exception as e:
                last_exception = e
                with allure.step(f"healing: visibility wait failed for {by}={sel}"):
                    _attach_screenshot(driver, name=f"healing_wait_failed_{desc}")
                continue
        # 4) Attempt generated fallback selectors: by id, by name, by xpath visible text attempts
        for attempt in range(2):
            with allure.step(f"healing: attempt generated fallback {attempt+1} for {desc}"):
                # attempt1: id from desc if plausible
                guess_id = desc.replace(" ", "_")
                try:
                    el = driver.find_element(By.ID, guess_id)
                    _attach_screenshot(driver, name=f"found_fallback_id_{desc}")
                    return el
                except Exception:
                    pass
                # attempt2: visible text xpath
                try:
                    xpath = f"//*[text()='{desc}']"
                    el = driver.find_element(By.XPATH, xpath)
                    _attach_screenshot(driver, name=f"found_fallback_text_{desc}")
                    return el
                except Exception:
                    pass
                try:
                    WebDriverWait(driver, 0.5).until(lambda d: True)
                except Exception:
                    pass
        # If all fail, raise
        raise last_exception if last_exception is not None else RuntimeError(f"Element {desc} not found")


def click_with_healing(driver, locator_candidates, desc):
    with allure.step(f"Click action: {desc}"):
        try:
            el = find_element_with_healing(driver, locator_candidates, desc)
            try:
                el.click()
                _attach_screenshot(driver, name=f"clicked_{desc}")
                return
            except Exception:
                # fallback to JS click immediately
                driver.execute_script("arguments[0].click();", el)
                _attach_screenshot(driver, name=f"clicked_js_primary_{desc}")
                return
        except Exception as first_exc:
            with allure.step(f"click failed: retrying once for {desc}"):
                _attach_screenshot(driver, name=f"click_failed_first_{desc}")
                try:
                    WebDriverWait(driver, 0.6).until(lambda d: True)
                except Exception:
                    pass
                try:
                    el = find_element_with_healing(driver, locator_candidates, desc)
                    try:
                        el.click()
                        _attach_screenshot(driver, name=f"clicked_retry_{desc}")
                        return
                    except Exception:
                        driver.execute_script("arguments[0].click();", el)
                        _attach_screenshot(driver, name=f"clicked_js_retry_{desc}")
                        return
                except Exception:
                    with allure.step(f"healing: fallback click for {desc}"):
                        _attach_screenshot(driver, name=f"click_failed_second_{desc}")
                        # Last attempt: execute JavaScript click if element present via wait
                        try:
                            el = WebDriverWait(driver, 3).until(EC.element_to_be_clickable(locator_candidates[0]))
                            driver.execute_script("arguments[0].click();", el)
                            _attach_screenshot(driver, name=f"clicked_js_{desc}")
                            return
                        except Exception as e:
                            _attach_screenshot(driver, name=f"click_all_failed_{desc}")
                            raise


def send_keys_with_healing(driver, locator_candidates, desc, value):
    with allure.step(f"Send keys: {desc} -> '{value}'"):
        try:
            el = find_element_with_healing(driver, locator_candidates, desc)
            try:
                el.clear()
            except Exception:
                pass
            el.send_keys(value)
            _attach_screenshot(driver, name=f"sentkeys_{desc}")
            return
        except Exception as first_exc:
            with allure.step(f"send_keys failed: retrying once for {desc}"):
                _attach_screenshot(driver, name=f"sendkeys_failed_first_{desc}")
                try:
                    WebDriverWait(driver, 0.6).until(lambda d: True)
                except Exception:
                    pass
                try:
                    el = find_element_with_healing(driver, locator_candidates, desc)
                    try:
                        el.clear()
                    except Exception:
                        pass
                    el.send_keys(value)
                    _attach_screenshot(driver, name=f"sentkeys_retry_{desc}")
                    return
                except Exception:
                    with allure.step(f"healing: fallback send_keys for {desc}"):
                        _attach_screenshot(driver, name=f"sendkeys_failed_second_{desc}")
                        # fallback: set value via JS
                        try:
                            el = find_element_with_healing(driver, locator_candidates, desc)
                            driver.execute_script("arguments[0].value = arguments[1];", el, value)
                            _attach_screenshot(driver, name=f"sentkeys_js_{desc}")
                            return
                        except Exception:
                            _attach_screenshot(driver, name=f"sendkeys_all_failed_{desc}")
                            raise


def helper_login_jan_sauce_demo_login_py(driver, username, password):
    with allure.step("POM jan_sauce_demo_login.py not available — using helper_login_jan_sauce_demo_login_py (auto-healed)"):
        allure.attach(json.dumps({"healed": "created helper_login_jan_sauce_demo_login_py"}), name="healing_info", attachment_type=AttachmentType.JSON)
        _attach_screenshot(driver, name="login_helper_start")
        send_keys_with_healing(driver, [(By.ID, "user-name"), (By.NAME, "user-name"), (By.XPATH, "//*[@id='user-name']")], "user-name", username)
        send_keys_with_healing(driver, [(By.ID, "password"), (By.NAME, "password"), (By.XPATH, "//*[@id='password']")], "password", password)
        click_with_healing(driver, [(By.ID, "login-button"), (By.NAME, "login-button"), (By.XPATH, "//*[@id='login-button']")], "login-button")


def helper_add_to_cart_jan_sauce_demo_login_py(driver):
    with allure.step("POM jan_sauce_demo_login.py not available — using helper_add_to_cart_jan_sauce_demo_login_py (auto-healed)"):
        allure.attach(json.dumps({"healed": "created helper_add_to_cart_jan_sauce_demo_login_py"}), name="healing_info", attachment_type=AttachmentType.JSON)
        _attach_screenshot(driver, name="add_to_cart_start")
        click_with_healing(driver, [(By.ID, "add-to-cart-sauce-labs-backpack"), (By.XPATH, "//*[@id='add-to-cart-sauce-labs-backpack']")], "add-to-cart-sauce-labs-backpack")
        _attach_screenshot(driver, name="added_to_cart")


def helper_open_cart_jan_sauce_demo_login_py(driver):
    with allure.step("POM jan_sauce_demo_login.py not available — using helper_open_cart_jan_sauce_demo_login_py (auto-healed)"):
        allure.attach(json.dumps({"healed": "created helper_open_cart_jan_sauce_demo_login_py"}), name="healing_info", attachment_type=AttachmentType.JSON)
        _attach_screenshot(driver, name="open_cart_start")
        click_with_healing(driver, [(By.CLASS_NAME, "shopping_cart_link"), (By.XPATH, "//a[@class='shopping_cart_link']"), (By.ID, "shopping_cart_container")], "shopping_cart_link")


def helper_checkout_fill_jan_sauce_demo_login_py(driver, first_name, last_name, postal_code):
    with allure.step("POM jan_sauce_demo_login.py not available — using helper_checkout_fill_jan_sauce_demo_login_py (auto-healed)"):
        allure.attach(json.dumps({"healed": "created helper_checkout_fill_jan_sauce_demo_login_py"}), name="healing_info", attachment_type=AttachmentType.JSON)
        click_with_healing(driver, [(By.ID, "checkout"), (By.XPATH, "//*[@id='checkout']")], "checkout")
        _attach_screenshot(driver, name="checkout_page")
        send_keys_with_healing(driver, [(By.ID, "firstName"), (By.NAME, "firstName"), (By.XPATH, "//*[@id='firstName']")], "firstName", first_name)
        send_keys_with_healing(driver, [(By.ID, "lastName"), (By.NAME, "lastName"), (By.XPATH, "//*[@id='lastName']")], "last_name", last_name)
        send_keys_with_healing(driver, [(By.ID, "postalCode"), (By.NAME, "postalCode"), (By.XPATH, "//*[@id='postalCode']")], "postalCode", postal_code)
        click_with_healing(driver, [(By.ID, "continue"), (By.XPATH, "//*[@id='continue']")], "continue")


def helper_finish_purchase_jan_sauce_demo_login_py(driver):
    with allure.step("POM jan_sauce_demo_login.py not available — using helper_finish_purchase_jan_sauce_demo_login_py (auto-healed)"):
        allure.attach(json.dumps({"healed": "created helper_finish_purchase_jan_sauce_demo_login_py"}), name="healing_info", attachment_type=AttachmentType.JSON)
        click_with_healing(driver, [(By.ID, "finish"), (By.XPATH, "//*[@id='finish']")], "finish")
        _attach_screenshot(driver, name="finished_purchase")
        click_with_healing(driver, [(By.ID, "back-to-products"), (By.XPATH, "//*[@id='back-to-products']")], "back-to-products")
        _attach_screenshot(driver, name="back_to_products_clicked")


@pytest.fixture(scope="function")
def driver():
    options = Options()
    options.add_argument("--start-maximized")
    driver = webdriver.Chrome(options=options)
    try:
        yield driver
    finally:
        try:
            driver.quit()
        except Exception:
            pass


def _handle_assertion_failure(driver, exc):
    with allure.step("Assertion failed - collecting artifacts"):
        try:
            allure.attach(driver.get_screenshot_as_png(), name="assertion_failure_screenshot", attachment_type=AttachmentType.PNG)
        except Exception:
            pass
        try:
            allure.attach(driver.page_source.encode("utf-8"), name="assertion_failure_page_source", attachment_type=AttachmentType.HTML)
        except Exception:
            pass
        reports_dir = _write_report_artifacts(driver, tag="assertion_failure")
        with open(os.path.join(reports_dir, "note.txt"), "w") as f:
            f.write(f"Artifacts saved in {reports_dir}\n")
        raise exc


@allure.feature("Purchase Flow")
@allure.story("TC01 - Successful purchase (standard_user)")
def test_TC01_successful_purchase_standard_user(driver):
    # Exactly one primary assertion at the end: user is back on inventory page after purchase
    with allure.step("Start test: TC01 - Successful purchase (standard_user)"):
        driver.get(BASE_URL)
        _attach_screenshot(driver, name="landing_page")
        try:
            if len(driver.window_handles) > 1:
                with allure.step("Switching to new window as recorded"):
                    switch_to_new_window(driver)
                    _attach_screenshot(driver, name="switched_to_new_window")
        except Exception:
            with allure.step("No new window to switch to or switch failed"):
                _attach_screenshot(driver, name="switch_new_window_failed")

    with allure.step("POM availability check and healing"):
        allure.attach(json.dumps({"note": "jan_sauce_demo_login.py methods present in source list but class name not provided; using helpers"}), name="healing_summary", attachment_type=AttachmentType.JSON)
        _attach_screenshot(driver, name="pom_healing")

    with allure.step("Login with standard_user"):
        try:
            helper_login_jan_sauce_demo_login_py(driver, "standard_user", "secret_sauce")
        except Exception as e:
            _attach_screenshot(driver, name="login_error")
            raise

    with allure.step("Add product to cart"):
        try:
            helper_add_to_cart_jan_sauce_demo_login_py(driver)
        except Exception as e:
            _attach_screenshot(driver, name="add_to_cart_error")
            raise

    with allure.step("Open cart"):
        try:
            helper_open_cart_jan_sauce_demo_login_py(driver)
        except Exception as e:
            _attach_screenshot(driver, name="open_cart_error")
            raise

    with allure.step("Checkout and fill information"):
        try:
            helper_checkout_fill_jan_sauce_demo_login_py(driver, "t", "t", "hg")
        except Exception as e:
            _attach_screenshot(driver, name="checkout_fill_error")
            raise

    with allure.step("Finish purchase and return to products"):
        try:
            helper_finish_purchase_jan_sauce_demo_login_py(driver)
        except Exception as e:
            _attach_screenshot(driver, name="finish_error")
            raise

    with allure.step("Primary assertion: verify we are back on inventory page (/inventory.html)"):
        try:
            WebDriverWait(driver, 5).until(lambda d: "/inventory.html" in d.current_url)
            current = driver.current_url
            _attach_screenshot(driver, name="final_page")
            assert current.endswith("/inventory.html"), f"expected to be on inventory.html, but was on {current}"
        except AssertionError as ae:
            _handle_assertion_failure(driver, ae)
        except Exception as e:
            _handle_assertion_failure(driver, AssertionError(f"Primary assertion failed due to exception: {e}"))